README.md
# PhaseIIBeam_v19

History
Version 1 was the primary version with a few errors.

1/2/2018 Version 2: Changed the number of bins, minimum and maximum limits of histogram in proton.mac.

1/2/2018 Version 3: Changed the type of fNbElectronsStrip1, fNbElectronsStrip2 in RunAction.hh from G4double to G4int.

22/2/2018 Version 4: Added two couples of BPIX modules (one in front of and one behind the DUT). Calculating the entrance point of the primary proton inside Pixel Detector 1 and its exit point from Pixel Detector 2.

23/2/2018 Version 5: Decreased the thickness of pixel modules. Calculating the deflection angle of the primary track between the two pixel modules and filling with it histogram 84.

1/3/2018 Version 6: Local electric field added to each strip sensor. Global electric field could be added through ElectricFieldSetup.

7/3/2018 Version 7: Positions of primary inside strip sensors 1 and 2 displayed.

9/3/2018 Version 8: Midpoints of primary inside each detector displayed.

13/3/2018 Version 9: Let A = fPointPix1mid, B = fPointDet1mid, C = fPointDet2mid, D = fPointPix2mid. Let dB = distance between B and AD and dC = distance between C and AD. Histograms filled with dB and dC for each event.

13/3/2018 Version 10: Let B' be the AD point with same z as B. Let C' be the AD point with the same z as C. Histograms filled with B'Bx, B'By, C'Cx, C'Cy for each event.

13/3/2018 Version 11: Bug in filling the histograms per strip fixed.

15/3/2018 Version 12: World material changed to air.

20/3/2018 Version 13: Kinetic energy of secondaries calculated and could be printed out.

Versions 14 and 15 were non-working and full of bugs.

11/4/2018 Version 16: Possibility for 2S DUT rotation about the x-axis added.

11/4/2018 Version 17: Strips a and b for each sensor separated.

12/4/2018 Version 18: Changed "Air" material. Added CO2 and ArgonGas to it.

3/5/2018 Version 19: Changed titles of histograms 119-122.


src/RunAction.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
/// \file electromagnetic/TestEm18/src/RunAction.cc
/// \brief Implementation of the RunAction class
//
// $Id: RunAction.cc 82401 2014-06-18 14:43:54Z gcosmo $
//
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include "RunAction.hh"
#include "DetectorConstruction.hh"
#include "PrimaryGeneratorAction.hh"
#include "HistoManager.hh"
#include "EventAction.hh"

#include "G4Event.hh"
#include "G4Run.hh"
#include "G4RunManager.hh"
#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"
#include "G4VisAttributes.hh"
#include "G4UImanager.hh"
#include "G4EmCalculator.hh"
#include "G4Step.hh"

#include "Randomize.hh"
#include <iomanip>

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

RunAction::RunAction(DetectorConstruction* det, PrimaryGeneratorAction* kin)
:G4UserRunAction(),fDetector(det), fPrimary(kin), fHistoManager(0)
{ 
  fHistoManager = new HistoManager(); 
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

RunAction::~RunAction()
{ 
  delete fHistoManager; 
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void RunAction::BeginOfRunAction(const G4Run*)
{
  //initialisation
  //
  fEnergyDeposit1 = 0.;
  fEnergyDeposit2 = 0.;
  
  fNbCharged1 = fNbCharged2 = fNbNeutral1 = fNbNeutral2 = 0;
  fNbPositronsB1 = fNbPositrons1 = fNbPositrons12 = fNbPositrons2 = fNbPositronsA2 = fNbElectronsB1 = fNbElectrons1 = fNbElectrons12 = fNbElectrons2 = fNbElectronsA2 = fNbPionsPlus1 = fNbPionsPlus2 = fNbPionsMinus1 = fNbPionsMinus2 = fNbMuonsPlus1 = fNbMuonsPlus2 = fNbMuonsMinus1 = fNbMuonsMinus2 = 0;
  fNbDelta12 = fNbDelta21 = 0;
  fEnergyCharged1 = fEnergyCharged2 = fEnergyNeutral1 = fEnergyNeutral2 = 0.;  
  fEmin1[0] = fEmin1[1] = DBL_MAX;
  fEmax1[0] = fEmax1[1] = 0.;    
  fEmin2[0] = fEmin2[1] = DBL_MAX;
  fEmax2[0] = fEmax2[1] = 0.;   

  for (G4int k = 0; k < 1017; k = k + 1) {
    fNbElectronsStrip1a[k] = 0;
    fNbElectronsStrip1b[k] = 0;
    fNbElectronsStrip2a[k] = 0;
    fNbElectronsStrip2b[k] = 0;
  }
    
  fNbSteps = 0;
  fTrackLength = fSecTrackLength = fTertTrackLength = fSecDetTrackLength = 0.;
   
  //histograms
  //
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  if ( analysisManager->IsActive() ) {
    analysisManager->OpenFile();
  }       

  // do not save Rndm status
  G4RunManager::GetRunManager()->SetRandomNumberStore(false);
  CLHEP::HepRandom::showEngineStatus();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void RunAction::EndOfRunAction(const G4Run* aRun)
{
  G4int nbEvents = aRun->GetNumberOfEvent();
  if (nbEvents == 0) return;
  
  G4Material* material = fDetector->GetMaterialD();
  G4double length  = fDetector->GetSizeW();
   
  G4ParticleDefinition* particle = fPrimary->GetParticleGun()
                                          ->GetParticleDefinition();
  G4String partName = particle->GetParticleName();
  G4double ePrimary = fPrimary->GetParticleGun()->GetParticleEnergy();
  
  G4int prec = G4cout.precision(3);
  G4cout << "\n ======================== run summary ======================\n";
  G4cout << "\n The run was " << nbEvents << " " << partName << " of "
         << G4BestUnit(ePrimary,"Energy") << " through " 
         << G4BestUnit(length,"Length") << ".";
  G4cout << "\n ===========================================================\n";
  G4cout << G4endl;
  
  //save histograms      
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  analysisManager->FillH1(32, fNbPositrons1);
  analysisManager->FillH1(33, fNbPositrons2);
  analysisManager->FillH1(34, fNbElectrons1);
  analysisManager->FillH1(35, fNbElectrons2);
  analysisManager->FillH1(36, fNbPionsPlus1);
  analysisManager->FillH1(37, fNbPionsPlus2);
  analysisManager->FillH1(38, fNbPionsMinus1);
  analysisManager->FillH1(39, fNbPionsMinus2);
  analysisManager->FillH1(40, fNbMuonsPlus1);
  analysisManager->FillH1(41, fNbMuonsPlus2);
  analysisManager->FillH1(42, fNbMuonsMinus1);
  analysisManager->FillH1(43, fNbMuonsMinus2);  
  analysisManager->FillH1(44, fNbDelta12);
  analysisManager->FillH1(45, fNbDelta21); 
  analysisManager->FillH1(46, fNbPositronsB1);
  analysisManager->FillH1(47, fNbElectronsB1);
  analysisManager->FillH1(48, fNbPositrons12);
  analysisManager->FillH1(49, fNbElectrons12);  
  analysisManager->FillH1(50, fNbPositronsA2);
  analysisManager->FillH1(51, fNbElectronsA2); 
  for (G4int i = 84; i < 92; i = i + 1) {
    analysisManager->FillH1(i, fNbElectronsStrip1a[i+421]);
  }
  for (G4int j = 92; j < 100; j = j + 1) {
    analysisManager->FillH1(j, fNbElectronsStrip1b[j+413]);
  }
  for (G4int k = 100; k < 108; k = k + 1) {
    analysisManager->FillH1(k, fNbElectronsStrip2a[k+405]);
  }
  for (G4int l = 108; l < 116; l = l + 1) {
    analysisManager->FillH1(l, fNbElectronsStrip2b[l+397]);
  }
  if ( analysisManager->IsActive() ) {
    analysisManager->Write();
    analysisManager->CloseFile();
  }      
    
  if (particle->GetPDGCharge() == 0.) return;
   
  G4cout.precision(5);
  
  //track length
  //
  G4double trackLPerEvent = fTrackLength/nbEvents;
  G4double trackLS = fSecTrackLength/nbEvents;
  G4double trackLT = fTertTrackLength/nbEvents;
  G4double nbStepPerEvent = double(fNbSteps)/nbEvents;
  G4double stepSize = fTrackLength/fNbSteps;
  G4double trackSecDet = fSecDetTrackLength/nbEvents;
  
  G4cout 
    << "\n TrackLength= " 
    << G4BestUnit(trackLPerEvent, "Length")
    << "\n Secondary TrackLength= " 
    << G4BestUnit(trackLS, "Length")
    << "\n Tertiary TrackLength= " 
    << G4BestUnit(trackLT, "Length")
    << "\t nb of steps = " << nbStepPerEvent
    << "  stepSize= " << G4BestUnit(stepSize, "Length")
    << "  stepSize for secondaries created in detectors = " << G4BestUnit(trackSecDet, "Length")
    << G4endl;
      
  //charged secondaries (ionization, direct pair production)
  //
  G4double energy1PerEvent = fEnergyCharged1/nbEvents;
  G4double energy2PerEvent = fEnergyCharged2/nbEvents;
  G4double nb1PerEvent = double(fNbCharged1)/nbEvents;
  G4double nb2PerEvent = double(fNbCharged2)/nbEvents;
  G4double meanEkin1 = 0.;
  if (fNbCharged1) meanEkin1 = fEnergyCharged1/fNbCharged1;
  G4double meanEkin2 = 0.;
  if (fNbCharged2) meanEkin2 = fEnergyCharged2/fNbCharged2;
  
  G4cout 
    << "\n d-rays  : eLoss1/primary= " 
    << G4BestUnit(energy1PerEvent, "Energy")
    << "\n d-rays  : eLoss2/primary= " 
    << G4BestUnit(energy2PerEvent, "Energy")
    << "\t  nb1 of d-rays= " << nb1PerEvent
    << "\t  nb2 of d-rays= " << nb2PerEvent
    << G4endl;
         
  //neutral secondaries (bremsstrahlung, pixe)
  //
  energy1PerEvent = fEnergyNeutral1/nbEvents;
  energy2PerEvent = fEnergyNeutral2/nbEvents;
  nb1PerEvent = double(fNbNeutral1)/nbEvents;
  nb2PerEvent = double(fNbNeutral2)/nbEvents;
  meanEkin1 = 0.;
  if (fNbNeutral1) meanEkin1 = fEnergyNeutral1/fNbNeutral1;
  meanEkin2 = 0.;
  if (fNbNeutral2) meanEkin2 = fEnergyNeutral2/fNbNeutral2;
  
  G4cout 
    << "\n gamma   : eLoss1/primary= " 
    << G4BestUnit(energy1PerEvent, "Energy")
    << "\n gamma   : eLoss2/primary= " 
    << G4BestUnit(energy2PerEvent, "Energy")
    << "\t  nb1 of gammas= " << nb1PerEvent
    << "\t  nb2 of gammas= " << nb2PerEvent
    << "  <Tkin1>= " << G4BestUnit(meanEkin1, "Energy")
    << "  Tmin1= "   << G4BestUnit(fEmin1[1],  "Energy")
    << "  Tmax1= "   << G4BestUnit(fEmax1[1],  "Energy")
    << "  <Tkin2>= " << G4BestUnit(meanEkin2, "Energy")
    << "  Tmin2= "   << G4BestUnit(fEmin2[1],  "Energy")
    << "  Tmax2= "   << G4BestUnit(fEmax2[1],  "Energy")
    << G4endl;
    

  G4EmCalculator emCal;
  
  //local energy deposit for Detector 1
  //
  energy1PerEvent = fEnergyDeposit1/nbEvents;
  //
  G4double r01  = emCal.GetRangeFromRestricteDEDX(ePrimary,particle,material);  
  G4double r11 = r01 - trackLPerEvent;
  G4double etry1 = ePrimary - energy1PerEvent;  
  G4double efinal1 = 0.;
  if (r11 > 0.) efinal1 = GetEnergyFromRestrictedRange(r11,particle,material,etry1);
  G4double dEtable1 = ePrimary - efinal1;
  G4double ratio1 = 0.;
  if (dEtable1 > 0.) ratio1 = energy1PerEvent/dEtable1;
    
  G4cout 
    << "\n deposit1 : eLoss1/primary= " 
    << G4BestUnit(energy1PerEvent, "Energy")
    << "\t <dEcut1 > table= " 
    << G4BestUnit(dEtable1, "Energy")
    << "   ---> simul1/reference= " << ratio1          
    << G4endl;
    
  //total energy transferred for Detector 1
  //
  G4double energyTotal1 = fEnergyDeposit1 + fEnergyCharged1 + fEnergyNeutral1;
  energy1PerEvent = energyTotal1/nbEvents;
  //
  r01  = emCal.GetCSDARange(ePrimary,particle,material);  
  r11 = r01 - trackLPerEvent;
  etry1 = ePrimary - energy1PerEvent;
  efinal1 = 0.;
  if (r11 > 0.) efinal1 = GetEnergyFromCSDARange(r11,particle,material,etry1);
  dEtable1 = ePrimary - efinal1;
  ratio1 = 0.;
  if (dEtable1 > 0.) ratio1 = energy1PerEvent/dEtable1;
    
  G4cout 
    << "\n total   : eLoss1/primary= " 
    << G4BestUnit(energy1PerEvent, "Energy")
    << "\t <dEfull1> table= " 
    << G4BestUnit(dEtable1, "Energy")
    << "   ---> simul1/reference= " << ratio1           
    << G4endl; 

  //local energy deposit for Detector 2
  //
  energy2PerEvent = fEnergyDeposit2/nbEvents;
  //
  G4double r02  = emCal.GetRangeFromRestricteDEDX(ePrimary,particle,material);  
  G4double r12 = r02 - trackLPerEvent;
  G4double etry2 = ePrimary - energy2PerEvent;  
  G4double efinal2 = 0.;
  if (r12 > 0.) efinal2 = GetEnergyFromRestrictedRange(r12,particle,material,etry2);
  G4double dEtable2 = ePrimary - efinal2;
  G4double ratio2 = 0.;
  if (dEtable2 > 0.) ratio2 = energy2PerEvent/dEtable2;
    
  G4cout 
    << "\n deposit2 : eLoss2/primary= " 
    << G4BestUnit(energy2PerEvent, "Energy")
    << "\t <dEcut2 > table= " 
    << G4BestUnit(dEtable2, "Energy")
    << "   ---> simul2/reference= " << ratio2          
    << G4endl;
    
  //total energy transferred for Detector 2
  //
  G4double energyTotal2 = fEnergyDeposit2 + fEnergyCharged2 + fEnergyNeutral2;
  energy2PerEvent = energyTotal2/nbEvents;
  //
  r02  = emCal.GetCSDARange(ePrimary,particle,material);  
  r12 = r02 - trackLPerEvent;
  etry2 = ePrimary - energy2PerEvent;
  efinal2 = 0.;
  if (r12 > 0.) efinal2 = GetEnergyFromCSDARange(r12,particle,material,etry2);
  dEtable2 = ePrimary - efinal2;
  ratio2 = 0.;
  if (dEtable2 > 0.) ratio2 = energy2PerEvent/dEtable2;
    
  G4cout 
    << "\n total   : eLoss2/primary= " 
    << G4BestUnit(energy2PerEvent, "Energy")
    << "\t <dEfull2> table= " 
    << G4BestUnit(dEtable2, "Energy")
    << "   ---> simul2/reference= " << ratio2           
    << G4endl; 

  G4cout.precision(prec);

  // show Rndm status
  CLHEP::HepRandom::showEngineStatus();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

G4double RunAction::GetEnergyFromRestrictedRange(G4double range,
            G4ParticleDefinition* particle, G4Material* material, G4double Etry)
{
  G4EmCalculator emCal;
    
  G4double Energy = Etry, dE = 0., dEdx;
  G4double r, dr;
  G4double err  = 1., errmax = 0.00001;
  G4int    iter = 0 , itermax = 10;  
  while (err > errmax && iter < itermax) {
    iter++;
    Energy -= dE;
    r = emCal.GetRangeFromRestricteDEDX(Energy,particle,material);
    dr = r - range;          
    dEdx = emCal.GetDEDX(Energy,particle,material);
    dE = dEdx*dr;
    err = std::abs(dE)/Energy;    
  }
  if (iter == itermax) {
    G4cout 
    << "\n  ---> warning: RunAction::GetEnergyFromRestRange() did not converge"
    << "   Etry = " << G4BestUnit(Etry,"Energy")
    << "   Energy = " << G4BestUnit(Energy,"Energy")
    << "   err = " << err
    << "   iter = " << iter << G4endl;
  }         
         
  return Energy;         
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

G4double RunAction::GetEnergyFromCSDARange(G4double range,
            G4ParticleDefinition* particle, G4Material* material, G4double Etry)
{
  G4EmCalculator emCal;
    
  G4double Energy = Etry, dE = 0., dEdx;
  G4double r, dr;
  G4double err  = 1., errmax = 0.00001;
  G4int    iter = 0 , itermax = 10;  
  while (err > errmax && iter < itermax) {
    iter++;
    Energy -= dE;
    r = emCal.GetCSDARange(Energy,particle,material);
    dr = r - range;          
    dEdx = emCal.ComputeTotalDEDX(Energy,particle,material);
    dE = dEdx*dr;
    err = std::abs(dE)/Energy;
  }
  if (iter == itermax) {
    G4cout 
    << "\n  ---> warning: RunAction::GetEnergyFromCSDARange() did not converge"
    << "   Etry = " << G4BestUnit(Etry,"Energy")
    << "   Energy = " << G4BestUnit(Energy,"Energy")
    << "   err = " << err
    << "   iter = " << iter << G4endl;
  }         
         
  return Energy;         
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......


src/HistoManager.cc
//
// ********************************************************************
// * License and Disclaimer                                           *
// *                                                                  *
// * The  Geant4 software  is  copyright of the Copyright Holders  of *
// * the Geant4 Collaboration.  It is provided  under  the terms  and *
// * conditions of the Geant4 Software License,  included in the file *
// * LICENSE and available at  http://cern.ch/geant4/license .  These *
// * include a list of copyright holders.                             *
// *                                                                  *
// * Neither the authors of this software system, nor their employing *
// * institutes,nor the agencies providing financial support for this *
// * work  make  any representation or  warranty, express or implied, *
// * regarding  this  software system or assume any liability for its *
// * use.  Please see the license in the file  LICENSE  and URL above *
// * for the full disclaimer and the limitation of liability.         *
// *                                                                  *
// * This  code  implementation is the result of  the  scientific and *
// * technical work of the GEANT4 collaboration.                      *
// * By using,  copying,  modifying or  distributing the software (or *
// * any work based  on the software)  you  agree  to acknowledge its *
// * use  in  resulting  scientific  publications,  and indicate your *
// * acceptance of all terms of the Geant4 Software license.          *
// ********************************************************************
//
//
// $Id: HistoManager.cc 72242 2013-07-12 08:44:19Z gcosmo $
// 
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

#include <CLHEP/Units/SystemOfUnits.h>
#include <sstream>

#include "HistoManager.hh"
#include "G4UnitsTable.hh"
#include "G4SystemOfUnits.hh"

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::HistoManager()
  :fFileName("phaseiiv1")
{
  Book();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

HistoManager::~HistoManager()
{
  delete G4AnalysisManager::Instance();
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......

void HistoManager::Book()
{
  // Create or get analysis manager
  // The choice of analysis technology is done via selection of a namespace
  // in HistoManager.hh
  // Creating a tree container to handle histograms and ntuples.
  // This tree is associated to an output file.
  //
  G4AnalysisManager* analysisManager = G4AnalysisManager::Instance();
  analysisManager->SetFileName(fFileName);
  analysisManager->SetVerboseLevel(1);
  analysisManager->SetActivation(true);   //enable inactivation of histograms

  
  // Define histograms start values
  // const G4int kNbofStrips = 1016;
  const G4int kMaxHisto = 123;

  const G4String id[] = { "0", "1", "2", "3" , "4", "5", "6", "7", "8", "9" , "10", "11", "12", "13", "14", "15", "16", "17", "18", "19", "20", "21", "22", "23", "24", "25", "26", "27", "28", "29", "30", "31", "32", "33", "34", "35", "36", "37", "38", "39", "40", "41", "42", "43", "44", "45", "46", "47", "48", "49", "50", "51", "52", "53", "54", "55", "56", "57", "58", "59", "60", "61", "62", "63", "64", "65", "66", "67", "68", "69", "70", "71", "72", "73", "74", "75", "76", "77", "78", "79", "80", "81", "82", "83", "84", "85", "86", "87", "88", "89", "90", "91", "92", "93", "94", "95", "96", "97", "98", "99", "100", "101", "102", "103", "104", "105", "106", "107", "108", "109", "110", "111", "112", "113", "114", "115", "116", "117", "118", "119", "120", "121", "122" };
  const G4String title[] =
                { "dummy",                                                       //0
                  "continuous energy loss along primary track for Detector 1",   //1
                  "continuous energy loss along primary track for Detector 2",   //2
                  "energy from secondaries for Detector 1",                      //3
                  "energy from secondaries for Detector 2",                      //4
                  "energy from tertiaries for Detector 1",                       //5
                  "energy from tertiaries for Detector 2",                       //6
                  "total energy lost by primary track for Detector 1",           //7
                  "total energy lost by primary track for Detector 2",           //8
                  "energy spectrum of secondary e-+ for Detector 1",             //9
                  "energy spectrum of secondary e-+ for Detector 2",             //10
                  "energy spectrum of secondary gamma for Detector 1",           //11
                  "energy spectrum of secondary gamma for Detector 2",           //12
                  "energy spectrum of tertiary e-+ for Detector 1",              //13
                  "energy spectrum of tertiary e-+ for Detector 2",              //14
                  "energy spectrum of tertiary gamma for Detector 1",            //15
                  "energy spectrum of tertiary gamma for Detector 2",            //16
                  "step size for primary",                                       //17
                  "step size for secondaries",                                   //18
                  "step size for tertiaries",                                    //19
                  "step size for secondaries created in detectors",              //20
                  "x-polarization of secondaries",                               //21
                  "y-polarization of secondaries",                               //22
                  "z-polarization of secondaries",                               //23
                  "x-polarization of tertiaries",                                //24
                  "y-polarization of tertiaries",                                //25
                  "z-polarization of tertiaries",                                //26
                  "track length for primary",                                    //27
                  "total track length of secondaries",                           //28
                  "total track length of tertiaries",                            //29
                  "total track length of secondaries created in detectors",      //30
                  "track length of secondaries created in detectors",            //31
                  "number of positrons created in Detector 1",       		 //32
                  "number of positrons created in Detector 2",       		 //33
                  "number of electrons created in Detector 1",       		 //34
                  "number of electrons created in Detector 2",       		 //35
                  "number of pi+ created in Detector 1",             		 //36
                  "number of pi+ created in Detector 2",             		 //37
                  "number of pi- created in Detector 1",              		 //38
                  "number of pi- created in Detector 2",             		 //39
                  "number of mu+ created in Detector 1",             		 //40
                  "number of mu+ created in Detector 2",             		 //41
                  "number of mu- created in Detector 1",             		 //42
                  "number of mu- created in Detector 2",              		 //43
		  "number of delta e- created in Detector 1 that have reached Detector 2", //44
		  "number of delta e- created in Detector 2 that have reached Detector 1", //45
                  "number of e+ created before Detector 1",               	 //46
                  "number of e- created before Detector 1",               	 //47
		  "number of e+ created between Detector 1 and Detector 2", 	 //48
		  "number of e- created between Detector 1 and Detector 2", 	 //49
                  "number of e+ created after Detector 2",               	 //50
                  "number of e- created after Detector 2",                	 //51
                  "continuous energy loss along primary track for strip 505 a Detector 1",   //52
                  "continuous energy loss along primary track for strip 506 a Detector 1",   //53
                  "continuous energy loss along primary track for strip 507 a Detector 1",   //54
                  "continuous energy loss along primary track for strip 508 a Detector 1",   //55
                  "continuous energy loss along primary track for strip 509 a Detector 1",   //56
                  "continuous energy loss along primary track for strip 510 a Detector 1",   //57
                  "continuous energy loss along primary track for strip 511 a Detector 1",   //58
                  "continuous energy loss along primary track for strip 512 a Detector 1",   //59
                  "continuous energy loss along primary track for strip 505 b Detector 1",   //60
                  "continuous energy loss along primary track for strip 506 b Detector 1",   //61
                  "continuous energy loss along primary track for strip 507 b Detector 1",   //62
                  "continuous energy loss along primary track for strip 508 b Detector 1",   //63
                  "continuous energy loss along primary track for strip 509 b Detector 1",   //64
                  "continuous energy loss along primary track for strip 510 b Detector 1",   //65
                  "continuous energy loss along primary track for strip 511 b Detector 1",   //66
                  "continuous energy loss along primary track for strip 512 b Detector 1",   //67
                  "continuous energy loss along primary track for strip 505 a Detector 2",   //68
                  "continuous energy loss along primary track for strip 506 a Detector 2",   //69
                  "continuous energy loss along primary track for strip 507 a Detector 2",   //70
                  "continuous energy loss along primary track for strip 508 a Detector 2",   //71
                  "continuous energy loss along primary track for strip 509 a Detector 2",   //72
                  "continuous energy loss along primary track for strip 510 a Detector 2",   //73
                  "continuous energy loss along primary track for strip 511 a Detector 2",   //74
                  "continuous energy loss along primary track for strip 512 a Detector 2",   //75
                  "continuous energy loss along primary track for strip 505 b Detector 2",   //76
                  "continuous energy loss along primary track for strip 506 b Detector 2",   //77
                  "continuous energy loss along primary track for strip 507 b Detector 2",   //78
                  "continuous energy loss along primary track for strip 508 b Detector 2",   //79
                  "continuous energy loss along primary track for strip 509 b Detector 2",   //80
                  "continuous energy loss along primary track for strip 510 b Detector 2",   //81
                  "continuous energy loss along primary track for strip 511 b Detector 2",   //82
                  "continuous energy loss along primary track for strip 512 b Detector 2",   //83
                  "number of electrons created in strip 505 a Detector 1",   //84
                  "number of electrons created in strip 506 a Detector 1",   //85
                  "number of electrons created in strip 507 a Detector 1",   //86
                  "number of electrons created in strip 508 a Detector 1",   //87
                  "number of electrons created in strip 509 a Detector 1",   //88
                  "number of electrons created in strip 510 a Detector 1",   //89
                  "number of electrons created in strip 511 a Detector 1",   //90
                  "number of electrons created in strip 512 a Detector 1",   //91
                  "number of electrons created in strip 505 b Detector 1",   //92
                  "number of electrons created in strip 506 b Detector 1",   //93
                  "number of electrons created in strip 507 b Detector 1",   //94
                  "number of electrons created in strip 508 b Detector 1",   //95
                  "number of electrons created in strip 509 b Detector 1",   //96
                  "number of electrons created in strip 510 b Detector 1",   //97
                  "number of electrons created in strip 511 b Detector 1",   //98
                  "number of electrons created in strip 512 b Detector 1",   //99
                  "number of electrons created in strip 505 a Detector 2",   //100
                  "number of electrons created in strip 506 a Detector 2",   //101
                  "number of electrons created in strip 507 a Detector 2",   //102
                  "number of electrons created in strip 508 a Detector 2",   //103
                  "number of electrons created in strip 509 a Detector 2",   //104
                  "number of electrons created in strip 510 a Detector 2",   //105
                  "number of electrons created in strip 511 a Detector 2",   //106
                  "number of electrons created in strip 512 a Detector 2",   //107
                  "number of electrons created in strip 505 b Detector 2",   //108
                  "number of electrons created in strip 506 b Detector 2",   //109
                  "number of electrons created in strip 507 b Detector 2",   //110
                  "number of electrons created in strip 508 b Detector 2",   //111
                  "number of electrons created in strip 509 b Detector 2",   //112
                  "number of electrons created in strip 510 b Detector 2",   //113
                  "number of electrons created in strip 511 b Detector 2",   //114
                  "number of electrons created in strip 512 b Detector 2",   //115
                  "deflection angle of primary particle (rad)",              //116
                  "distance between midpoint of primary inside Detector 1 and AD",   //117
                  "distance between midpoint of primary inside Detector 2 and AD",   //118
                  "2S first-plane residuals in x-direction",   //119
                  "2S first-plane residuals in y-direction",   //120
                  "2S second-plane residuals in x-direction",   //121
                  "2S second-plane residuals in y-direction"    //122
                  //"midpoint of primary particle inside Pixel Detector 1",
                  //"midpoint of primary particle inside Strip Detector 1",
                  //"midpoint of primary particle inside Strip Detector 2",
                  //"midpoint of primary particle inside Pixel Detector 2"
                 };
            
  // Default values (to be reset via /analysis/h1/set command)               
  G4int nbins = 100;
  G4double vmin = -1.5;
  G4double vmax = 100.;

  // Create all histograms as inactivated 
  // as we have not yet set nbins, vmin, vmax
  for (G4int k=0; k<kMaxHisto; k++) {
    G4int ih = analysisManager->CreateH1(id[k], title[k], nbins, vmin, vmax);
    analysisManager->SetH1Activation(ih, false);
  }

  //Create H3 histograms
  //G4int ih1 = analysisManager->CreateH3("85", "midpoint of primary particle inside Pixel Detector 1", 10000, -2*cm, 2*cm, 10000, -2*cm, 2*cm, 10000, -2*cm, 2*cm, "cm", "cm", "cm");
  //analysisManager->SetH3Activation(ih1, false);
  //G4int ih2 = analysisManager->CreateH3("86", "midpoint of primary particle inside Strip Detector 1", 1000, -2*mm, 2*mm, 1000, -2*mm, 2*mm, 1000, -2*mm, 2*mm, "mm", "mm", "mm");
  //analysisManager->SetH3Activation(ih2, false);
  //G4int ih3 = analysisManager->CreateH3("87", "midpoint of primary particle inside Strip Detector 2", 1000, -2*mm, 2*mm, 1000, -2*mm, 2*mm, 1000, -2*mm, 2*mm, "mm", "mm", "mm");
  //analysisManager->SetH3Activation(ih3, false);
  //G4int ih4 = analysisManager->CreateH3("88", "midpoint of primary particle inside Pixel Detector 2", 10000, -2*cm, 2*cm, 10000, -2*cm, 2*cm, 10000, -2*cm, 2*cm, "cm", "cm", "cm");
  //analysisManager->SetH3Activation(ih4, false);

  // nTuples, column l: 1 to 1016 correspond to Detector 1, 2 to 2032 correspond to Detector 2
  //
  analysisManager->SetNtupleDirectoryName("ntuple");
  analysisManager->SetFirstNtupleId(1);       
  analysisManager->CreateNtuple("1", "Primary Particle Tuple 1");
  for (G4int l=0; l<1017; l++) {
     std::ostringstream os1;
     os1 <<"EdepStripDetector1a_" << l;
     analysisManager->CreateNtupleDColumn(os1.str());          //
  }
  for (G4int m1=0; m1<1017; m1++) {
     std::ostringstream os2;
     os2 <<"EdepStripDetector1b_" << m1;
     analysisManager->CreateNtupleDColumn(os2.str());          //
  }
  for (G4int ma=0; ma<1017; ma++) {
     std::ostringstream os3;
     os3 <<"EdepStripDetector2a_" << ma;
     analysisManager->CreateNtupleDColumn(os3.str());          //
  }
  for (G4int mb=0; mb<1017; mb++) {
     std::ostringstream os4;
     os4 <<"EdepStripDetector2b_" << mb;
     analysisManager->CreateNtupleDColumn(os4.str());          //
  }
  //for (G4int n=0; n<1017; n++) {
     //std::ostringstream os3;
     //os3 <<"NbElectronsStripDetector1_" << n;
     //analysisManager->CreateNtupleDColumn(os3.str());          //
  //}
  //for (G4int o=0; o<1017; o++) {
     //std::ostringstream os4;
     //os4 <<"NbElectronsStripDetector2_" << o;
     //analysisManager->CreateNtupleDColumn(os4.str());          //
  //}
  analysisManager->FinishNtuple();
  
  //analysisManager->SetNtupleActivation(false);  
}

//....oooOO0OOooo........oooOO0OOooo........oooOO0OOooo........oooOO0OOooo......
